### PR TITLE
RPC protocol update + multi-view plumbing [ready]

### PR DESCRIPTION
**Pull request by cmyr**
~edit: After a bit of playing around there may be a few performance regressions I want to look at, but this is good for review.~ (all good)

This is approximately where I would like it to be in terms of the API modifications discussed in #191. It also includes a variety of naming changes, which are intended to more clearly document the use of particular APIs & methods. I'm touchy about renaming other people's code: please consider all naming changes to be suggestions, and I'm happy to revert them. 😄

### RPC Changes:
Because this already involved breaking changes to the RPC protocol, I took the liberty of normalizing certain other RPC signatures.

- In *all* RPC methods, the word `tab` as an identifier of 'client viewports' has been replaced with `view_id`.
- in *all* RPC methods, the word `filename` has been replaced with the word `file_path`. (We were using both in different places, I just grepped the src directory and picked the most common).
- `TabCommand` is now `CoreCommand`. (not a protocol change)
-  `NewTab` and `DeleteTab` (`new_tab`, `delete_tab`) have been removed from `CoreCommand`.
- `NewView { file_path }`, `CloseView { view_id }` and `Save { view_id, file_path }` have been added to `CoreCommand`.
- `Open` and `Save` have been removed from `EditCommand`.

In error messages and other names, the word `tab` has been replaced with `core`, where appropriate.

### Other changes:

The other main change here is that `Editor` now behaves as it if has multiple views, although the actual multi-view support isn't implemented. To reduce the diff size, `Editor` still does everything through a `self.view` instance. There is code in place to dynamically swap this variable out when an event arrives associated with a new view, but these codepaths are not currently available.

Anyway: the real consequence of this is that events and responses are associated with a specific view_id, rather than a buffer+view (e.g. a 'tab'). This plus the clarified view and editor lifecycle stuff enabled by the RPC changes set me up to _hopefully_ start implementing provisional plugin support. Maybe? 🤞 

### To think about:

- What is the role of `TabCtx`, now? with message routing now based on `view_id`, which isn't owned by a particular `TabCtx`, this should be rethought.
- How will actual multi-view support happen, and what refactoring is required to make that possible? It still feels like there's a lot of state split between the editor and the view. I haven't looked into this very much.
- **Are  there any other protocol changes to make now**, while we're already breaking things?

### CODE REVIEW

**Comment by cmyr** [comment]
@raphlinus 🚀

File: rust/core-lib/src/tabs.rs [hunk 1]
```diff
@@ -15,112 +15,238 @@
...
+        self.id_counter += 1;
+        format!("buffer-id-{}", self.id_counter)
+    }
+
+    pub fn do_rpc(&mut self, cmd: CoreCommand, rpc_peer: &MainPeer<W>) -> Option<Value> {
+        use rpc::CoreCommand::*;
+
+        match cmd {
+            CloseView { view_id } => {
+                self.do_close_view(view_id); self.
```
Comment (raphlinus) [inline rust/core-lib/src/tabs.rs]:
```
error: no field `None` on type `&mut tabs::Tabs<W>`
  --> core-lib/src/tabs.rs:94:51
   |
94 |                   self.do_close_view(view_id); self.
   |  ___________________________________________________^ starting here...
95 | |                 None
   | |____________________^ ...ending here

error: aborting due to previous error
```

**Comment by cmyr** [comment]
I was testing out RLS vim integration. `self.` produces lots of useful completion suggestions! 😞

**Comment by raphlinus** [comment]
Thanks!

**Pull request merged**